File: ios/Muraka/Core/Common/SessionEvents.swift
import Foundation

/// Something that happened to the session which the interface has to react to.
enum SessionEvent: Sendable {
    /// The refresh token was rejected. The session is over.
    ///
    /// The contributor is returned to sign-in and **the queue is kept** — scenario 6 of
    /// `mobile-shared/sync-protocol.md`. Clearing it here would throw away reef data because
    /// a token expired, which is not a reason to lose anything.
    case refreshFailed
    /// An admin suspended the account. Sign out with an explanation, not a silent bounce.
    case accountDisabled
}

/// A bus for the two events that can end a session from underneath the interface.
///
/// It exists because the place that discovers them — a token refresh deep inside a
/// background upload — has no way to reach a navigation controller, and passing one down
/// there would be far worse than a stream.
///
/// `AsyncStream` with a buffering policy rather than an unbounded one: emission must never
/// block the uploader on an interface that may not be listening.
final class SessionEvents: Sendable {
    private let continuation: AsyncStream<SessionEvent>.Continuation
    let events: AsyncStream<SessionEvent>

    init() {
        var capturedContinuation: AsyncStream<SessionEvent>.Continuation?
        events = AsyncStream(bufferingPolicy: .bufferingNewest(1)) { capturedContinuation = $0 }
        // The stream's initialiser runs its closure synchronously, so this is always set.
        continuation = capturedContinuation ?? AsyncStream<SessionEvent>.makeStream().continuation
    }

    func send(_ event: SessionEvent) {
        continuation.yield(event)
    }
}

File: android/core/common/src/main/kotlin/mv/muraka/core/common/SessionEvents.kt
package mv.muraka.core.common

import kotlinx.coroutines.channels.BufferOverflow
import kotlinx.coroutines.flow.MutableSharedFlow
import kotlinx.coroutines.flow.SharedFlow
import kotlinx.coroutines.flow.asSharedFlow
import javax.inject.Inject
import javax.inject.Singleton

/** Something that happened to the session, which the UI has to react to. */
sealed interface SessionEvent {
    /**
     * The refresh token was rejected. The session is over.
     *
     * The contributor is returned to sign-in and **the queue is kept** - scenario 6 of
     * `mobile-shared/sync-protocol.md`. Clearing it here would throw away reef data
     * because a token expired, which is not a reason to lose anything.
     */
    data object RefreshFailed : SessionEvent

    /** An admin suspended the account. Sign out with an explanation, not a silent bounce. */
    data object AccountDisabled : SessionEvent
}

/**
 * A bus for the two events that can end a session from underneath the UI.
 *
 * It exists because the place that discovers them - an OkHttp `Authenticator`, deep
 * inside a background upload - has no way to reach a navigation controller, and passing
 * one down there would be far worse than a flow.
 *
 * `extraBufferCapacity` with `DROP_OLDEST`: emission must never suspend, because the
 * authenticator is on a network thread that cannot afford to block on a UI that may not
 * be collecting.
 */
@Singleton
class SessionEvents @Inject constructor() {
    private val _events = MutableSharedFlow<SessionEvent>(
        extraBufferCapacity = 1,
        onBufferOverflow = BufferOverflow.DROP_OLDEST,
    )
    val events: SharedFlow<SessionEvent> = _events.asSharedFlow()

    fun emit(event: SessionEvent) {
        _events.tryEmit(event)
    }
}
